# Imports and Load Data

In [47]:
# ── IMPORTS ───────────────────────────────────────────────────────
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from dotenv import load_dotenv
from sqlalchemy import create_engine
import os
from scipy.stats import chi2_contingency
from itertools import combinations

# load environment variables — database password stored in .env
load_dotenv()

# connect to recruiting_analytics database
engine = create_engine(
    f"postgresql://postgres:{os.getenv('DB_PASSWORD')}@localhost:5432/recruiting_analytics"
)


In [48]:
# ── LOAD RAW DATA FROM POSTGRESQL ─────────────────────────────────

# jobs and candidates are small — load normally
df_jobs       = pd.read_sql('SELECT * FROM raw.jobs',       engine)
df_candidates = pd.read_sql('SELECT * FROM raw.candidates', engine)

# applications is large — read in chunks to avoid memory issues
# then combine into one dataframe
df_applications = pd.concat(
    pd.read_sql('SELECT * FROM raw.applications', engine, chunksize=5000)
)

print(f"jobs:         {len(df_jobs)} rows")
print(f"candidates:   {len(df_candidates)} rows")
print(f"applications: {len(df_applications)} rows")

jobs:         600 rows
candidates:   23000 rows
applications: 29019 rows


# Data Cleaning

In [49]:
# ── CLEAN SOURCE LABELS ───────────────────────────────────────────

# first let's see what messy values we have
print("Raw source values:")
print(df_applications['source'].value_counts())

Raw source values:
source
LinkedIn              9986
Company Website       5273
Referral              3809
Indeed                2933
Recruiter Outreach    2006
linkedin               772
LinkedIn Jobs          749
Linkedin               742
Career Page            538
company website        529
referral               388
Employee Referral      344
indeed                 298
Indeed.com             281
Outreach               191
Sourced                180
Name: count, dtype: int64


In [50]:
# ── STANDARDIZE SOURCE LABELS ─────────────────────────────────────

# mapping of all messy variants to their canonical label
source_map = {
    # LinkedIn variants
    'LinkedIn':      'LinkedIn',
    'linkedin':      'LinkedIn',
    'LinkedIn Jobs': 'LinkedIn',
    'Linkedin':      'LinkedIn',
    
    # Company Website variants
    'Company Website': 'Company Website',
    'company website': 'Company Website',
    'Career Page':     'Company Website',
    
    # Referral variants
    'Referral':          'Referral',
    'referral':          'Referral',
    'Employee Referral': 'Referral',
    
    # Indeed variants
    'Indeed':     'Indeed',
    'indeed':     'Indeed',
    'Indeed.com': 'Indeed',
    
    # Recruiter Outreach variants
    'Recruiter Outreach': 'Recruiter Outreach',
    'Outreach':           'Recruiter Outreach',
    'Sourced':            'Recruiter Outreach',
}

# apply the mapping
df_applications['source'] = df_applications['source'].map(source_map)

# verify — should only see 5 clean values now
print("Cleaned source values:")
print(df_applications['source'].value_counts())

# check for any unmapped values (would show as NaN)
print(f"\nUnmapped sources: {df_applications['source'].isna().sum()}")

Cleaned source values:
source
LinkedIn              12249
Company Website        6340
Referral               4541
Indeed                 3512
Recruiter Outreach     2377
Name: count, dtype: int64

Unmapped sources: 0


In [51]:
# ── FIX OUT OF ORDER DATES ────────────────────────────────────────

# find cases where screen_date is before applied_date
# these were introduced as messiness in generation
out_of_order = (
    df_applications['screen_date'].notna() & 
    (df_applications['screen_date'] < df_applications['applied_date'])
)

print(f"Out of order dates found: {out_of_order.sum()}")

# fix by swapping the dates back
df_applications.loc[out_of_order, ['applied_date', 'screen_date']] = \
    df_applications.loc[out_of_order, ['screen_date', 'applied_date']].values

# verify fix
still_broken = (
    df_applications['screen_date'].notna() & 
    (df_applications['screen_date'] < df_applications['applied_date'])
)
print(f"Out of order dates remaining: {still_broken.sum()}")

Out of order dates found: 0
Out of order dates remaining: 0


In [52]:
# convert date columns to datetime
date_cols = ['applied_date', 'screen_date', 'interview_date', 
             'final_round_date', 'offer_date', 'hire_date']

for col in date_cols:
    df_applications[col] = pd.to_datetime(df_applications[col])

# now calculate days to hire
df_applications['days_to_hire'] = (
    df_applications['hire_date'] - df_applications['applied_date']
).dt.days

print(f"Hired candidates with days_to_hire: {df_applications['days_to_hire'].notna().sum()}")
print(f"\nDays to hire summary:")
print(df_applications['days_to_hire'].describe())

Hired candidates with days_to_hire: 303

Days to hire summary:
count    303.000000
mean      30.039604
std        9.535490
min       12.000000
25%       23.000000
50%       28.000000
75%       37.000000
max       54.000000
Name: days_to_hire, dtype: float64


In [53]:
# days to hire by department — engineering should be longest
hired = df_applications[df_applications['outcome'] == 'Hired'].merge(
    df_jobs[['job_id', 'department']], on='job_id'
)

print(hired.groupby('department')['days_to_hire'].mean().sort_values(ascending=False).round(1))

department
Engineering           42.5
Data & Analytics      31.2
Product               30.1
HR                    24.9
Finance               24.9
Customer Success      24.1
Operations            23.8
Legal & Compliance    22.3
Marketing             22.3
Sales                 17.0
Name: days_to_hire, dtype: float64


In [54]:
# ── MISSING REJECTION REASONS ─────────────────────────────────────

# check how many rejected candidates have no rejection reason
rejected = df_applications[df_applications['outcome'] == 'Rejected']
missing = rejected['rejection_reason'].isna().sum()
print(f"Rejected candidates missing reason: {missing} ({missing/len(rejected)*100:.1f}%)")

# flag them as 'Not Provided' — more honest than leaving blank
# this is a common real-world cleaning decision
df_applications.loc[
    (df_applications['outcome'] == 'Rejected') & 
    (df_applications['rejection_reason'].isna()),
    'rejection_reason'
] = 'Not Provided'

# verify
print(f"Missing after fix: {df_applications['rejection_reason'].isna().sum()}")

Rejected candidates missing reason: 3856 (15.2%)
Missing after fix: 3602


In [55]:
# check who the remaining nulls belong to
print(df_applications[df_applications['rejection_reason'].isna()]['outcome'].value_counts())
print(f"\nTotal nulls: {df_applications['rejection_reason'].isna().sum()}")

outcome
Active      2876
Withdrew     423
Hired        303
Name: count, dtype: int64

Total nulls: 3602


In [56]:
# check for any other issues worth flagging
print(df_applications.dtypes)
print(f"\nMissing values per column:")
print(df_applications.isna().sum())

application_id                str
candidate_id                  str
job_id                        str
source                        str
current_stage                 str
outcome                       str
rejection_reason              str
applied_date        datetime64[s]
screen_date         datetime64[s]
interview_date      datetime64[s]
final_round_date    datetime64[s]
offer_date          datetime64[s]
hire_date           datetime64[s]
days_to_hire              float64
dtype: object

Missing values per column:
application_id          0
candidate_id            0
job_id                  0
source                  0
current_stage           0
outcome                 0
rejection_reason     3602
applied_date            0
screen_date         26116
interview_date      27294
final_round_date    28072
offer_date          28484
hire_date           28716
days_to_hire        28716
dtype: int64


In [57]:
# convert days_to_hire to integer — no partial days
df_applications['days_to_hire'] = df_applications['days_to_hire'].astype('Int64')
# Int64 (capital I) handles NaN values, regular int64 doesn't

In [58]:
# ── WRITE CLEANED DATA TO POSTGRESQL ─────────────────────────────

# jobs and candidates have no cleaning needed — copy as-is to clean schema
df_jobs.to_sql('jobs', engine, schema='clean', if_exists='append', index=False)
print(f"clean.jobs loaded: {len(df_jobs)} rows")

df_candidates.to_sql('candidates', engine, schema='clean', if_exists='append', index=False)
print(f"clean.candidates loaded: {len(df_candidates)} rows")

# applications — cleaned version with standardized sources and days_to_hire
df_applications.to_sql('applications', engine, schema='clean', if_exists='append', index=False)
print(f"clean.applications loaded: {len(df_applications)} rows")

print("\nAll clean data loaded ✅")

clean.jobs loaded: 600 rows


clean.candidates loaded: 23000 rows
clean.applications loaded: 29019 rows

All clean data loaded ✅


# Hypothesis Testing

In [59]:
df_applications['outcome'].value_counts()

outcome
Rejected    25417
Active       2876
Withdrew      423
Hired         303
Name: count, dtype: int64

In [60]:
# ── HYPOTHESIS TESTING ────────────────────────────────────────────
# Question: does candidate source affect hire rate?
# H0 (null): all sources have equal likelihood of resulting in a hire
# H1 (alternative): at least one source has a significantly different hire rate

# filter to completed applications only — exclude Active candidates
# Active = still in process, outcome unknown
completed = df_applications[df_applications['outcome'] != 'Active'].copy().reset_index(drop=True)

print(f"Total applications: {len(df_applications)}")
print(f"Completed applications: {len(completed)}")

# build contingency table — rows are sources, columns are hired/not hired
contingency = pd.crosstab(
    completed['source'],
    completed['outcome'] == 'Hired'
)
contingency.columns = ['Not Hired', 'Hired']

print("\nContingency table:")
print(contingency)
print(f"\nHire rate by source:")
print((contingency['Hired'] / contingency.sum(axis=1)).round(4))

Total applications: 29019
Completed applications: 26143

Contingency table:
                    Not Hired  Hired
source                              
Company Website          5689     36
Indeed                   3153     27
LinkedIn                10955     66
Recruiter Outreach       2089     55
Referral                 3954    119

Hire rate by source:
source
Company Website       0.0063
Indeed                0.0085
LinkedIn              0.0060
Recruiter Outreach    0.0257
Referral              0.0292
dtype: float64


In [61]:
# ── CHI-SQUARE TEST ───────────────────────────────────────────────

chi2, p_value, dof, expected = chi2_contingency(contingency)

print(f"Chi-square statistic: {chi2:.4f}")
print(f"Degrees of freedom:   {dof}")
print(f"P-value:              {p_value:.6f}")
print(f"\nSignificance level: 0.05")
print(f"Result: {'REJECT null hypothesis ✅' if p_value < 0.05 else 'FAIL to reject null hypothesis'}")
print(f"\nInterpretation:")
if p_value < 0.05:
    print("There is statistically significant evidence that candidate source")
    print("affects hire rate. Not all sources perform equally.")
else:
    print("There is insufficient evidence that candidate source affects hire rate.")

# check expected frequencies — all should be >= 5 for chi-square to be valid
print(f"\nMinimum expected frequency: {expected.min():.1f} (should be >= 5)")

Chi-square statistic: 194.3805
Degrees of freedom:   4
P-value:              0.000000

Significance level: 0.05
Result: REJECT null hypothesis ✅

Interpretation:
There is statistically significant evidence that candidate source
affects hire rate. Not all sources perform equally.

Minimum expected frequency: 24.8 (should be >= 5)


In [62]:
# ── PAIRWISE CHI-SQUARE TESTS WITH BONFERRONI CORRECTION ─────────

# define pairs to test — all combinations of sources
sources = contingency.index.tolist()
pairs = list(combinations(sources, 2))

# bonferroni correction — divide significance threshold by number of tests
n_tests = len(pairs)
bonferroni_threshold = 0.05 / n_tests

print(f"Number of pairwise tests: {n_tests}")
print(f"Bonferroni corrected threshold: {bonferroni_threshold:.4f}")
print(f"\n{'Source 1':<25} {'Source 2':<25} {'Chi2':>8} {'P-value':>12} {'Significant':>12}")
print("-" * 85)

results = []

for source1, source2 in pairs:
    # build 2x2 contingency table for this pair
    pair_table = contingency.loc[[source1, source2]]
    
    # run chi-square test
    chi2, p, dof, expected = chi2_contingency(pair_table)
    
    # is it significant after bonferroni correction?
    significant = p < bonferroni_threshold
    
    print(f"{source1:<25} {source2:<25} {chi2:>8.2f} {p:>12.6f} {'✅' if significant else '❌':>12}")
    
    results.append({
        'source_1':    source1,
        'source_2':    source2,
        'chi2':        chi2,
        'p_value':     p,
        'significant': significant
    })

df_pairwise = pd.DataFrame(results)

print(f"\nSignificant pairs (p < {bonferroni_threshold:.4f}):")
print(df_pairwise[df_pairwise['significant']][['source_1', 'source_2', 'p_value']])

Number of pairwise tests: 10
Bonferroni corrected threshold: 0.0050

Source 1                  Source 2                      Chi2      P-value  Significant
-------------------------------------------------------------------------------------
Company Website           Indeed                        1.12     0.290886            ❌
Company Website           LinkedIn                      0.02     0.895223            ❌
Company Website           Recruiter Outreach           49.49     0.000000            ✅
Company Website           Referral                     78.89     0.000000            ✅
Indeed                    LinkedIn                      2.01     0.156718            ❌
Indeed                    Recruiter Outreach           23.75     0.000001            ✅
Indeed                    Referral                     37.85     0.000000            ✅
LinkedIn                  Recruiter Outreach           74.07     0.000000            ✅
LinkedIn                  Referral                    130.63  

## Hypothesis Test Conclusion

**Question:** Does candidate source affect hire rate?

**Overall test:** Rejected null hypothesis (p < 0.0001) — sources are not equally likely to result in a hire.

**Pairwise findings (Bonferroni corrected, threshold p < 0.005):**
- Referral (2.92% hire rate) and Recruiter Outreach (2.57% hire rate) significantly outperform LinkedIn, Indeed, and Company Website
- Referral and Recruiter Outreach are not significantly different from each other
- LinkedIn, Indeed, and Company Website are not significantly different from each other

**Business recommendation:** Invest in referral programs and proactive recruiter outreach. Job board postings (LinkedIn, Indeed, Company Website) perform similarly and significantly worse than pre-vetted candidate channels.

In [63]:
# ── RECRUITER EFFICIENCY BY SOURCE ───────────────────────────────
# how many applications must be processed per hire for each source?
# lower = more efficient (less recruiter time per hire)

apps_per_hire = (contingency.sum(axis=1) / contingency['Hired']).round(1)
apps_per_hire = apps_per_hire.sort_values()

print("Applications processed per hire by source:")
print("-" * 45)
for source, count in apps_per_hire.items():
    print(f"  {source:<25} {count:>6.1f} applications/hire")

# how much less efficient is LinkedIn vs Referral?
ratio = apps_per_hire['LinkedIn'] / apps_per_hire['Referral']
print(f"\nLinkedIn requires {ratio:.1f}x more applications per hire than Referral")
print(f"— a significant recruiter time cost at scale")

Applications processed per hire by source:
---------------------------------------------
  Referral                    34.2 applications/hire
  Recruiter Outreach          39.0 applications/hire
  Indeed                     117.8 applications/hire
  Company Website            159.0 applications/hire
  LinkedIn                   167.0 applications/hire

LinkedIn requires 4.9x more applications per hire than Referral
— a significant recruiter time cost at scale
